runtime/vm/zcinterp.m4

```diff
@@ -153,70 +153,101 @@ END_CURRENT
 dnl    code to be 2 with the below CL(G)FI instruction indicating a transitent 
 dnl    transaction abort.
 dnl
 dnl 2. TX was a constrained transaction:
-dnl    If this was the case then we are safe to execute the read barrier because
-dnl    the JIT has ensured that there will be no side-effects of the transaction
-dnl    being aborted prior to reaching the guarded storage event handler. In this
-dnl    case it is functionally safe to execute the read barrier.
+dnl    This should not happen because we cannot handle constrained transactions
+dnl    in the hardware read barrier case. Because executing any of the following
+dnl    code will be functionally incorrect we force a crash here so we fail fast
+dnl    and can manually debug.
 dnl
-dnl See eclipse/openj9#2545 and eclipse/openj9#2790 for more details.
-    TM J9TR_VMThread_gsParameters_GSECI(J9VMTHREAD),128
-    JZ LABEL_NAME(L_GS_CALLHELPER)
-    TM J9TR_VMThread_gsParameters_GSECI(J9VMTHREAD),64
-    JNZ LABEL_NAME(L_GS_CALLHELPER)
-    CLFI_GPR J9SP,0
-    J LABEL_NAME(L_GS_DONE)
-PLACE_LABEL(L_GS_CALLHELPER)
-    RELOAD_SSP_SAVE_VALUE
-    STM_GPR r0,LAST_REG,JIT_GPR_SAVE_SLOT(0)
-    SWAP_SSP_VALUE
-    SAVE_HIWORD_REGS
-    LOAD_CAA
-    HANDLE_GS_EVENT_HELPER($1)
-PLACE_LABEL(L_GS_DONE)
-dnl Branch back to the instruction that triggered guarded storage event.
-    BRANCH_INDIRECT_ON_CONDITION(15,J9TR_VMThread_gsParameters_returnAddr,0,J9VMTHREAD)
-END_CURRENT
-})
+dnl See the following PRs for more details:
+dnl eclipse/openj9#2545
+dnl eclipse/openj9#2790
+dnl eclipse/openj9#3763
+        TM J9TR_VMThread_gsParameters_GSECI(J9VMTHREAD),128
+        JZ LABEL_NAME(L_GS_CALL_HELPER)
+        TM J9TR_VMThread_gsParameters_GSECI(J9VMTHREAD),64
+        JNZ LABEL_NAME(L_GS_FORCE_CRASH)
+        CLFI_GPR J9SP,0
+        J LABEL_NAME(L_GS_SKIP_HELPER)
 
-define({HANDLE_SOFTWARE_READ_BARRIER},{
-BEGIN_HELPER($1)
-    HANDLE_GS_EVENT_HELPER($1)
-dnl Branch back to the instruction that called software read barrier hanlder.
-    BR r14
-END_CURRENT
-})
+    PLACE_LABEL(L_GS_CALL_HELPER)
+        ST_GPR J9SP,J9TR_VMThread_sp(J9VMTHREAD)
+        LR_GPR CARG1,J9VMTHREAD
+        L_GPR CRA,J9TR_VMThread_javaVM(J9VMTHREAD)
+        L_GPR CRA,J9TR_JavaVM_invokeJ9ReadBarrier(CRA)
+        CALL_INDIRECT(CRA)
+
+dnl Load the updated object pointer into R9
+        LG r8,J9TR_VMThread_gsParameters_operandAddr(J9VMTHREAD)
+    ifdef({ASM_J9VM_GC_COMPRESSED_POINTERS},{
+        LLGF r8,0(r8)
+        LG r9,J9TR_VMThread_javaVM(J9VMTHREAD)
+        LG r9,J9TR_JavaVM_compressedPointersShift(r9)
+        SLLG r9,r8,0(r9)
+    },{
+        LG r9,0(r8)
+    })
+
+dnl Extract the target address register
+        LG r8,J9TR_VMThread_gsParameters_instructionAddr(J9VMTHREAD)
+        LLGH r8,0(r8)
+        NILF r8,240
+
+dnl We combine a shift right by 4 and a multiply by 8 to a shift
+dnl right by 1 to get the byte offset into the save area where to
+dnl store the updated object pointer
+        SRLG r8,r8,1
```

Comment:
Minor optimization: You can combine the `NILF r8, 240` and `SRLG r8,r8,1` into a single `RISBGN R8,R8,56,59,63` (might have to double check the indices calculation), which has a nicer bypass into the subsequent STG memref.

Reply:
I must admit this is mostly laziness on my part since encoding an `RISBGN` requires hand crafting a m4 macro which generates the hex encoding for the instruction. However you are completely correct. I'll take care of this tomorrow.

Reply:
Fixed in f95c0c2